packages/entity-server/src/routes/hierarchy/EntityRelationsRoute.ts

```diff
@@ -0,0 +1,150 @@
+  async buildDescendantAncestorPairs(descendants: EntityType[], ancestorType: string) {
+    const { hierarchyId } = this.req.ctx;
+    const descendantAncestorMapping = await this.req.models.entity.fetchAncestorDetailsByDescendantCode(
+      descendants.map(descendant => descendant.code),
+      hierarchyId,
+      ancestorType,
+    );
+
+    return Object.entries(descendantAncestorMapping).map(([descendant, { code: ancestor }]) => ({
+      descendant,
+      ancestor,
+    }));
+  }
+
+  async buildFormattedEntitiesByCode(ancestors: EntityType[], descendants: EntityType[]) {
+    const allEntities = [...ancestors, ...descendants];
+    const formattedEntities = await this.req.ctx.formatEntitiesForResponse(allEntities);
+    const formattedEntitiesByCode: Record<string, EntityResponseObject> = {};
+    allEntities.forEach((unformattedEntity, index) => {
+      formattedEntitiesByCode[unformattedEntity.code] = formattedEntities[index];
+    });
+    return formattedEntitiesByCode;
+  }
+
+  async groupByAncestorResponse(
+    descendants: EntityType[],
+    ancestorType: string,
+    flattenToCode: boolean,
+  ) {
+    const { entity, hierarchyId, allowedCountries } = this.req.ctx;
+
+    const ancestorCodeToDescendantCodes = reduceToArrayDictionary(
+      await this.buildDescendantAncestorPairs(descendants, ancestorType),
+      'ancestor',
+      'descendant',
+    );
+
+    if (flattenToCode) {
+      return ancestorCodeToDescendantCodes;
+    }
+
+    const ancestorsWithDescendantsCodes = Object.keys(ancestorCodeToDescendantCodes);
+    const ancestors =
+      entity.type === ancestorType
+        ? [entity]
+        : await entity.getDescendants(hierarchyId, {
+            code: ancestorsWithDescendantsCodes,
+            country_code: allowedCountries,
+          });
+
+    const formattedEntitiesByCode = await this.buildFormattedEntitiesByCode(ancestors, descendants);
+    return ancestors.map(ancestor => ({
+      ...formattedEntitiesByCode[ancestor.code],
+      descendants: ancestorCodeToDescendantCodes[ancestor.code].map(
+        descendantCode => formattedEntitiesByCode[descendantCode],
+      ),
+    }));
+  }
+
+  async groupByDescendantResponse(
+    descendants: EntityType[],
+    ancestorType: string,
+    flattenToCode: boolean,
+  ) {
+    const { entity, hierarchyId, allowedCountries } = this.req.ctx;
+
+    const descendantCodeToAncestorCode = reduceToDictionary(
+      await this.buildDescendantAncestorPairs(descendants, ancestorType),
+      'descendant',
+      'ancestor',
+    );
+
+    if (flattenToCode) {
+      return descendantCodeToAncestorCode;
+    }
+
+    const ancestorsWithDescendantsCodes = Object.values(descendantCodeToAncestorCode);
+    const ancestors =
+      entity.type === ancestorType
+        ? [entity]
+        : await entity.getDescendants(hierarchyId, {
+            code: ancestorsWithDescendantsCodes,
+            country_code: allowedCountries,
+          });
```

Comment:
This logic is the same in `groupByAncestorResponse`, so you could probably extract it to a `getRequestedAncestors` method or similar